File: Problems/0071. (M) Simplify Path/Python3/solution.py
class Solution:
  def simplifyPath(self, path: str) -> str:
    # stack

    dirs = []
    for t in path.split('/'):
      if not t or t == '.':
        continue
      if t == '..':
        if dirs:
          dirs.pop()
      else:
        dirs.append(t)
      
    return "/" + "/".join(dirs)

File: Problems/0071. (M) Simplify Path/Java/solution.java
class Solution {
  public String simplifyPath(String path) {
    // stack

    Deque<String> dirs = new ArrayDeque();
    String[] tokens = path.split("/");
    for (String t: tokens) {
      if (t.length() == 0 || t.equals("."))
        continue;
      if (t.equals("..")) {
        if (dirs.size() > 0)
          dirs.pop();
      }
      else
        dirs.push(t);
    }
    String ret = "";
    while (dirs.size() > 0)
      ret += "/" + dirs.removeLast();
    return ret.length() > 0 ? ret : "/";
  }
}
